src/compartment.c

```diff
@@ -239,22 +253,86 @@ void
 comp_register_ddc(struct Compartment* new_comp)
 {
     void* __capability new_ddc = cheri_address_set(manager_ddc, new_comp->base);
-    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size + new_comp->scratch_mem_stack_size);
+    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size);
     // TODO bounds double-check
     new_comp->ddc = new_ddc;
 }
 
+/* For a given Compartment `new_comp`, an address `intercept_target` pointing
+ * to a function found within the compartment which we would like to intercept,
+ * and a `intercept_data` struct representing information to perform the
+ * intercept, synthesize and inject intructions at the call point of the
+ * `intercept_target`, in order to perform a transition out of the compartment
+ * to call the appropriate function with higher privileges.
+ */
 void
 comp_add_intercept(struct Compartment* new_comp, uintptr_t intercept_target, struct func_intercept intercept_data)
 {
+    // TODO check whether negative values break anything in all these generated functions
     printf("Found `%s` func at %p.\n", intercept_data.func_name, (void*) intercept_target);
-    int offset = (intercept_data.redirect_func - intercept_target) / 4;
-    assert(offset < (1 << 27));
-    offset &= (1 << 26) - 1;
-    const int arm_b_instr_mask = 0b101 << 26;
-    uint32_t instr_binary = arm_b_instr_mask | offset;
-    int* init_addr_ptr = (void*) intercept_target;
-    struct intercept_patch new_patch = { init_addr_ptr, instr_binary };
+    int32_t new_instrs[INTERCEPT_INSTR_COUNT];
+    size_t new_instr_idx = 0;
+    uintptr_t comp_manager_cap_addr = new_comp->manager_caps + new_comp->active_manager_caps_count * sizeof(void* __capability); // TODO
+
+    const int32_t arm_function_target_register = 0b01010; // use `x10` for now
+    const int32_t arm_transition_target_register = 0b01011; // use `x11` for now
```

Comment:
Why write these in binary?

Reply:
Well yes, in that everything in C is just a human-readable representation and is translated to binary at some point, but `10` is much more obviously related to `x10` than `0b01010`. The two are equivalent in terms of the generated code.

I find binary literals most handy when describing things with bitfields (like [Morello instructions](https://github.com/Linaro/vixl/blob/morello/src/aarch64/constants-aarch64.h#L1921-L1927))